### PR TITLE
feat: add availability shown in rr reassign

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?
CleanShot 2024-10-24 at 09 51 21

Loom: https://www.loom.com/share/0931754f5d344b3ea830d17604c8f4b3

Reassigning to an unavailable host  (Admin only)
CleanShot 2024-10-24 at 10 38 38

## How should this be tested?

1) Login an book a busy event with one account - easier on a personal event tyoe
2) create a RR event type with a few member assigned
3) Book a meeting for the same time. 
4) Login as an team admin and reassign. You will notice you can assign to everyone regardless if they are busy or not.
5) Login as a team member. You will notice you can only assign to those in your team that are free. 

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: apps/web/components/booking/BookingListItem.tsx [hunk 1]
```diff
@@ -429,12 +429,14 @@ function BookingListItem(booking: BookingItemProps) {
         setIsOpenDialog={setIsOpenRescheduleDialog}
         bookingUId={booking.uid}
       />
-      <ReassignDialog
-        isOpenDialog={isOpenReassignDialog}
-        setIsOpenDialog={setIsOpenReassignDialog}
-        bookingId={booking.id}
-        teamId={booking.eventType?.team?.id || 0}
-      />
+      {isOpenReassignDialog && (
```
Comment (sean-brydon) [inline apps/web/components/booking/BookingListItem.tsx:432]: Conidtionally render this hook so all the data fetching doesn't happen till we open up the modal.

File: apps/web/components/dialog/ReassignDialog.tsx [hunk 2]
```diff
@@ -150,6 +153,24 @@ export const ReassignDialog = ({ isOpenDialog, setIsOpenDialog, teamId, bookingI
                 onChange={(event) => {
                   form.setValue("teamMemberId", event?.value);
                 }}
+                components={{
+                  Option: ({ children, ...props }) => {
+                    const status = props.data?.status;
```
Comment (sean-brydon) [inline apps/web/components/dialog/ReassignDialog.tsx:158]: Override the current options component to provide a way to pass a status dot into the select

File: packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts [hunk 3]
```diff
@@ -35,27 +52,49 @@ export const getRoundRobinHostsToReassign = async ({ ctx, input }: GetRoundRobin
...
+    gettingRoundRobinHostsToReassignLogger
+  );
+
+  const availableUserIds = new Set(availableUsers.map((u) => u.id));
+
+  const canViewReassignBusyUsers = isOrgAdmin || (await isTeamAdmin(prisma, ctx.user.id, eventType.teamId));
 
-  const hosts = eventTypeHosts.map((host) => host.user);
+  const roundRobinHostsToReassign = availableEventTypeUsers.reduce((acc, host) => {
+    const status = availableUserIds.has(host.id) ? "available" : "unavailable";
```
Comment (sean-brydon) [inline packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts:85]: If the available users set we created has the userId of the host we know they are available if not they are busy.

File: packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts [hunk 4]
```diff
@@ -35,27 +52,49 @@ export const getRoundRobinHostsToReassign = async ({ ctx, input }: GetRoundRobin
...
+    if (canViewReassignBusyUsers || status === "available") {
```
Comment (sean-brydon) [inline packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts:86]: Easy way to check if the user is an admin else - we only allow the available users into the final list

File: apps/web/components/booking/BookingListItem.tsx (hunk 1 above)
Comment (emrysal) [reply apps/web/components/booking/BookingListItem.tsx:432]: yeah boi!

File: packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts [hunk 5]
```diff
@@ -35,27 +52,49 @@ export const getRoundRobinHostsToReassign = async ({ ctx, input }: GetRoundRobin
...
+      dateTo: dayjs(booking.endTime).format(),
+      timeZone: eventType.timeZone || "UTC",
     },
-  });
+    gettingRoundRobinHostsToReassignLogger
+  );
+
+  const availableUserIds = new Set(availableUsers.map((u) => u.id));
+
+  const canViewReassignBusyUsers = isOrgAdmin || (await isTeamAdmin(prisma, ctx.user.id, eventType.teamId));
```
Comment (alishaz-polymath) [inline packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts:82]: We should add one for system-wide admin, thinking about our self-hosting folks!

**Review by alishaz-polymath — CHANGES_REQUESTED** [review]
Looks great, with a minor blocker

**Review by alishaz-polymath — APPROVED** [review]

File: packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts (hunk 5 above)
Comment (sean-brydon) [reply packages/trpc/server/routers/viewer/teams/roundRobin/getRoundRobinHostsToReasign.handler.ts:82]: As discussed will create a seperate PR down the line to make this a api endpoint for reassigning as these handlers only acessable via UI atm

**Pull request merged**